daml-lf/spec/daml-lf-1.rst

```diff
@@ -3556,7 +3557,238 @@ as described by the ledger model::
+
+     cid ∈ dom(st₀)
+     st₀(cid) = (Mod':T, vₜ, 'active')
+     'tpl' (x' : T) ↦ { …,'implements' Mod:I {…}, …, }  ∈ 〚Ξ〛Mod'
+     vᵢ = 'to_interface' @Mod:I @Mod':T vₜ
+     v₂ vᵢ  ⇓  Ok 'True'
+     'interface' (x : I) ↦ { …, 'choices' { …,
+        'choice' ChKind Ch (y : 'ContractId' Mod:I) (z : τ) : σ 'by' eₚ 'observers' eₒ ↦ eₐ, …
+        } } ∈ 〚Ξ〛Mod
+     eₚ[x ↦ vᵢ, z ↦ v₁]  ⇓  Ok vₚ
+     eₒ[x ↦ vᵢ, z ↦ v₁]  ⇓  Err E
+   —————————————————————————————————————————————————————————————————————— EvUpdExercIfaceObserversErr
+     'exercise_interface' @Mod:I Ch cid v₁ v₂ ‖ (st₀, keys₀)
+       ⇓ᵤ
+     (Err (Fatal "Choice observer evaluation failed"), ε)
+
+     cid ∈ dom(st₀)
+     st₀(cid) = (Mod':T, vₜ, 'active')
+     'tpl' (x' : T) ↦ { …,'implements' Mod:I {…}, …, }  ∈ 〚Ξ〛Mod'
+     vᵢ = 'to_interface' @Mod:I @Mod':T vₜ
+     v₂ vᵢ  ⇓  Ok 'True'
+     'interface' (x : I) ↦ { …, 'choices' { …,
+        'choice' ChKind Ch (y : 'ContractId' Mod:I) (z : τ) : σ 'by' eₚ 'observers' eₒ ↦ eₐ, …
+        } } ∈ 〚Ξ〛Mod
+     eₚ[x ↦ vᵢ, z ↦ v₁]  ⇓  Ok vₚ
+     eₒ[x ↦ vᵢ, z ↦ v₁]  ⇓  Ok vₒ
+     |v₁| > 100
+   —————————————————————————————————————————————————————————————————————— EvUpdExercIfaceNestingArgErr
+     'exercise_interface' @Mod:I Ch cid v₁ v₂ ‖ (st₀, keys₀)
+       ⇓ᵤ
+     (Err (Fatal "Value exceeds maximum nesting value"), ε)
+
+     cid ∈ dom(st₀)
+     st₀(cid) = (Mod':T, vₜ, 'active')
+     'tpl' (x' : T) ↦ { …,'implements' Mod:I {…}, …, }  ∈ 〚Ξ〛Mod'
+     vᵢ = 'to_interface' @Mod:I @Mod':T vₜ
+     v₂ vᵢ  ⇓  Ok 'True'
+     'interface' (x : I) ↦ { …, 'choices' { …,
+        'choice' ChKind Ch (y : 'ContractId' Mod:I) (z : τ) : σ 'by' eₚ 'observers' eₒ ↦ eₐ, …
+        } } ∈ 〚Ξ〛Mod
+     eₚ[x ↦ vᵢ, z ↦ v₁]  ⇓  Ok vₚ
+     eₒ[x ↦ vᵢ, z ↦ v₁]  ⇓  Ok vₒ
+     |v₁| ≤ 100
+     eₐ[x ↦ vᵢ, y ↦ cid, z ↦ v₁]  ⇓  Err E
+   —————————————————————————————————————————————————————————————————————— EvUpdExercIfaceBodyEvalErr
+     'exercise_interface' @Mod:I Ch cid v₁ v₂ ‖ (st₀, keys₀)
+       ⇓ᵤ
+     (Err E, 'exercise' vₚ (cid, Mod:T, vₜ) ChKind ε)
+
+     cid ∈ dom(st₀)
+     st₀(cid) = (Mod':T, vₜ, 'active')
+     'tpl' (x' : T) ↦ { …,'implements' Mod:I {…}, …, }  ∈ 〚Ξ〛Mod'
+     vᵢ = 'to_interface' @Mod:I @Mod':T vₜ
+     v₂ vᵢ  ⇓  Ok 'True'
+     'interface' (x : I) ↦ { …, 'choices' { …,
+        'choice' 'consuming' Ch (y : 'ContractId' Mod:I) (z : τ) : σ 'by' eₚ 'observers' eₒ ↦ eₐ, …
+        } } ∈ 〚Ξ〛Mod
+     eₚ[x ↦ vᵢ, z ↦ v₁]  ⇓  Ok vₚ
+     eₒ[x ↦ vᵢ, z ↦ v₁]  ⇓  Ok vₒ
+     |v₁| ≤ 100
+     eₐ[x ↦ vᵢ, y ↦ cid, z ↦ v₁]  ⇓  Ok uₐ
+     keys₁ = keys₀ - keys₀⁻¹(cid)
+     st₁ = st₀[cid ↦ (Mod:T, vₜ, 'inactive')]
+     uₐ ‖ (st₁, keys₁)  ⇓ᵤ  (Err E, tr)
+   —————————————————————————————————————————————————————————————————————— EvUpdExercIfaceConsumErr
+     'exercise_interface' @Mod:I Ch cid v₁ v₂ ‖ (st₀, keys₀)
+       ⇓ᵤ
+     (Err E, 'exercise' vₚ (cid, Mod:T, vₜ) 'consuming' tr)
```

Comment:
```suggestion
     (Err E, 'exercise' vₚ (cid, Mod:T, vₜ) 'consuming' tr)
```
I am a bit confuse why we have ε in the previous rule and not in this one.

Reply:
Because in the previous rule it's an error (possibly an exception) that is raised during evaluation of the update expression of a body choice, so no partial transaction is possible. In this rule we take into account that the update interpretation might result in a partial transaction. Btw this is the same for regular 'exercise'.

Reply:
make sens. Thanks. 

>  Btw this is the same for regular 'exercise'.
I know, I was wondering if it was not a problem there as well.